micro-rdk/src/common/data_collector.rs

```diff
@@ -0,0 +1,352 @@
+/// of the list of "capture_methods" in the "attributes" section of a
+/// component's configuration JSON object as stored in app. Each element
+/// of "capture_methods" is meant to produce an instance of `DataCollector`
+/// as defined below
+#[derive(Debug, Clone)]
+pub struct DataCollectorConfig {
+    pub method: CollectionMethod,
+    pub capture_frequency_hz: f32,
+}
+
+impl TryFrom<&Kind> for DataCollectorConfig {
+    type Error = AttributeError;
+    fn try_from(value: &Kind) -> Result<Self, Self::Error> {
+        if !value.contains_key("method")? {
+            return Err(AttributeError::KeyNotFound("method".to_string()));
+        }
+        if !value.contains_key("capture_frequency_hz")? {
+            return Err(AttributeError::KeyNotFound(
+                "capture_frequency_hz".to_string(),
+            ));
+        }
+        let method_str: String = value
+            .get("method")?
+            .ok_or(AttributeError::KeyNotFound("method".to_string()))?
+            .try_into()?;
+        let capture_frequency_hz = value
+            .get("capture_frequency_hz")?
+            .ok_or(AttributeError::KeyNotFound(
+                "capture_frequency_hz".to_string(),
+            ))?
+            .try_into()?;
+        // TODO: Collectors that take arguments (ex. Board Analogs)
+        // let parameters: &Kind = if let Ok(Some(params)) = value.get("additional_params") {
+        //     params
+        // } else {
+        //     &Kind::NullValue(0)
+        // };
+        let method = match method_str.as_str() {
+            "Readings" => CollectionMethod::Readings,
+            "AngularVelocity" => CollectionMethod::AngularVelocity,
+            "LinearAcceleration" => CollectionMethod::LinearAcceleration,
+            "LinearVelocity" => CollectionMethod::LinearVelocity,
+            // TODO: Power Sensor and Board collectors
+            // "Voltage" => CollectionMethod::Voltage,
+            // "Current" => CollectionMethod::Current,
+            // "Analogs" => {
+            //     let analog_reader_name: String = parameters.get("reader_name")?.ok_or(AttributeError::KeyNotFound("reader_name".to_string()))?.try_into()?;
+            //     CollectionMethod::Analogs(analog_reader_name)
+            // },
+            _ => {
+                return Err(AttributeError::ConversionImpossibleError);
+            }
+        };
+        Ok(DataCollectorConfig {
+            method,
+            capture_frequency_hz,
+        })
+    }
+}
+
+/// A CollectionMethod is an enum whose values are associated with
+/// a method on one or more component traits
+#[derive(Debug, Clone)]
+pub enum CollectionMethod {
+    Readings,
+    // MovementSensor methods
+    AngularVelocity,
+    LinearAcceleration,
+    LinearVelocity,
+    // TODO: PowerSensor methods
+    // Voltage,
+    // Current,
+    // TODO: Board
+    // Analogs(String)
+}
+
+impl CollectionMethod {
+    fn method_str(&self) -> String {
```

Comment:
If the only real use of this is for `fmt`, maybe just inline it in there rather than synthesizing a new method name?

Reply:
And anyway don't you get `ToString` for free if you implement `Display`?

Reply:
both fair points